awc/src/middleware/redirect.rs

```diff
@@ -0,0 +1,267 @@
+where
+    S: Service<ConnectRequest, Response = ConnectResponse, Error = SendRequestError> + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = RedirectServiceFuture<S>;
+
+    actix_service::forward_ready!(connector);
+
+    fn call(&self, req: ConnectRequest) -> Self::Future {
+        match req {
+            ConnectRequest::Tunnel(head, addr) => {
+                let fut = self.connector.call(ConnectRequest::Tunnel(head, addr));
+                RedirectServiceFuture::Tunnel { fut }
+            }
+            ConnectRequest::Client(head, body, addr) => {
+                let connector = self.connector.clone();
+                let max_redirect_times = self.max_redirect_times;
+
+                // backup the uri for reuse schema and authority.
+                let uri = match head {
+                    RequestHeadType::Owned(ref head) => head.uri.clone(),
+                    RequestHeadType::Rc(ref head, ..) => head.uri.clone(),
+                };
+
+                let fut = connector.call(ConnectRequest::Client(head, body, addr));
+
+                RedirectServiceFuture::Client {
+                    fut,
+                    max_redirect_times,
+                    uri: Some(uri),
+                    addr,
+                    connector: Some(connector),
+                }
+            }
+        }
+    }
+}
+
+pin_project_lite::pin_project! {
+    #[project = RedirectServiceProj]
+    pub enum RedirectServiceFuture<S>
+    where
+        S: Service<ConnectRequest, Response = ConnectResponse, Error = SendRequestError>,
+        S: 'static
+    {
+        Tunnel { #[pin] fut: S::Future },
+        Client {
+            #[pin]
+            fut: S::Future,
+            max_redirect_times: u8,
+            uri: Option<Uri>,
+            addr: Option<SocketAddr>,
+            connector: Option<Rc<S>>
+        }
+    }
+}
+
+impl<S> Future for RedirectServiceFuture<S>
+where
+    S: Service<ConnectRequest, Response = ConnectResponse, Error = SendRequestError> + 'static,
+{
+    type Output = Result<ConnectResponse, SendRequestError>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.as_mut().project() {
+            RedirectServiceProj::Tunnel { fut } => fut.poll(cx),
+            RedirectServiceProj::Client {
+                fut,
+                max_redirect_times,
+                uri,
+                addr,
+                connector,
+            } => {
+                match ready!(fut.poll(cx))? {
+                    ConnectResponse::Client(res) => match res.head().status {
+                        StatusCode::MOVED_PERMANENTLY
+                        | StatusCode::FOUND
+                        | StatusCode::SEE_OTHER
+                        | StatusCode::TEMPORARY_REDIRECT
+                        | StatusCode::PERMANENT_REDIRECT
+                            if *max_redirect_times > 0 =>
```

Comment:
Could be simplified to single match
 
```
ConnectResponse::Client(res) if res.head().status.is_redirection() && *max_redirect_times > 0 => {
```

Reply:
I'm not familiar with redirect but it seems there should be different treatment on all these status codes.(303, 307 for example) 

Your suggestion is a good one but I left them group together because I have not figured out details like how to reuse body for different cases.

Reply:
Added special handle for 303 and 307